internals/plan/plan.go

```diff
@@ -42,11 +68,90 @@ const (
 	defaultCheckThreshold = 3
 )
 
+var (
+	// sectionExtensions keeps a map of registered extensions.
+	sectionExtensions = map[string]SectionExtension{}
+
+	// sectionExtensionsOrder records the order in which the extensions were registered.
+	sectionExtensionsOrder = []string{}
+)
+
+// builtinSections represents all the built-in layer sections. This list is used
+// for identifying built-in fields in this package. It is unit tested to match
+// the YAML fields exposed in the Layer type, to catch inconsistencies.
+var builtinSections = []string{"summary", "description", "services", "checks", "log-targets"}
+
+// RegisterExtension adds a plan schema extension. All registrations must be
+// done before the plan library is used. The order in which extensions are
+// registered determines the order in which the sections are marshalled.
+// Extension sections are marshalled after the built-in sections.
+func RegisterExtension(field string, ext SectionExtension) {
+	if slices.Contains(builtinSections, field) {
+		panic(fmt.Sprintf("internal error: extension %q already used as built-in field", field))
+	}
+	if _, ok := sectionExtensions[field]; ok {
+		panic(fmt.Sprintf("internal error: extension %q already registered", field))
+	}
+	sectionExtensions[field] = ext
+	sectionExtensionsOrder = append(sectionExtensionsOrder, field)
+}
+
+// UnregisterExtension removes a plan schema extension. This is only
+// intended for use by tests during cleanup.
+func UnregisterExtension(field string) {
+	delete(sectionExtensions, field)
+	sectionExtensionsOrder = slices.DeleteFunc(sectionExtensionsOrder, func(n string) bool {
+		return n == field
+	})
+}
+
 type Plan struct {
 	Layers     []*Layer              `yaml:"-"`
 	Services   map[string]*Service   `yaml:"services,omitempty"`
 	Checks     map[string]*Check     `yaml:"checks,omitempty"`
 	LogTargets map[string]*LogTarget `yaml:"log-targets,omitempty"`
+
+	Sections map[string]Section `yaml:",inline"`
+}
+
+// MarshalYAML implements an override for top level omitempty tags handling.
+// This is required since Sections are based on an inlined map, for which
+// omitempty and inline together is not currently supported.
+func (p *Plan) MarshalYAML() (interface{}, error) {
```

Comment:
@benhoyt This new implementation of Marshal now insures the ordering of top level sections according to the discussed proposal. Let me know what you think.

Reply:
This is definitely an interesting implementation, but I'm guessing there's no way other than `reflect` to tell the YAML package to order the fields (a map doesn't have order or maybe uses alpha order, and a regular struct is static). I didn't even know you could do this (build a brand new struct type dynamically).

Reply:
If we only had the known built-in keys to work with (and then this was not required in the first place), it would have been possible to simply do something like:

```
func (p *Plan) MarshalYAML() (interface{}, error) {
    // Inline type to redefine ordering
    layout := &struct {
        A interface{} `yaml:"check,omitempty"`
        B interface{} `yaml:"logtargets,omitempty"`
        C interface{} `yaml:"services,omitempty"`
    } {
        A: p.Checks,
        B: p.LogTargets,
        C: p.Services,
    }
    return layout
}
```

A/B/C above can be defined with the proper name and type as well, but since YAML processing is reflection magic in any case, you can go crazy as above, and it will still work perfectly.

However, the issue is simply we do not know at runtime which extensions were added, so we need dynamic creation of some type for which order is still maintained, i.e struct.

Reply:
I believe this is the only way, but since we can improve this at any point in time, and Marshalling is only something we currently do on request from the API (and given YAML processing is reflection based in any case), I believe this is fine.

Resolving for now.